sdk/core/Azure.Core/src/Shared/ConnectionString.cs

```diff
@@ -0,0 +1,176 @@
+
+        public static ConnectionString Parse(string connectionString, string segmentSeparator = ";", string keyValueSeparator = "=", bool allowEmptyValues = false)
+        {
+            Validate(connectionString, segmentSeparator, keyValueSeparator, allowEmptyValues);
+            return new ConnectionString(ParseSegments(connectionString, segmentSeparator, keyValueSeparator), segmentSeparator, keyValueSeparator);
+        }
+
+        private ConnectionString(Dictionary<string, string> pairs, string pairSeparator, string keyValueSeparator)
+        {
+            _pairs = pairs;
+            _pairSeparator = pairSeparator;
+            _keyValueSeparator = keyValueSeparator;
+        }
+
+        public string GetRequired(string key) =>
+            _pairs.TryGetValue(key, out var value) ? value : throw new InvalidOperationException($"Required key '{key}' is missing in connection string.");
+
+        public string GetNonRequired(string key) =>
+            _pairs.TryGetValue(key, out var value) ? value : null;
+
+        public void Replace(string key, string value)
+        {
+            if (_pairs.ContainsKey(key))
+            {
+                _pairs[key] = value;
+            }
+        }
+
+        public override string ToString()
+        {
+            var stringBuilder = new StringBuilder();
+            var isFirst = true;
+            foreach (KeyValuePair<string, string> segment in _pairs)
+            {
+                if (isFirst)
+                {
+                    isFirst = false;
+                }
+                else
+                {
+                    stringBuilder.Append(_pairSeparator);
+                }
+
+                stringBuilder.Append(segment.Key);
+                if (segment.Value != null)
+                {
+                    stringBuilder.Append(_keyValueSeparator).Append(segment.Value);
+                }
+            }
+
+            return stringBuilder.ToString();
+        }
+
+
+        private static Dictionary<string, string> ParseSegments(in string connectionString, in string segmentSeparator, in string keyValueSeparator)
+        {
+            var segments = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            var segmentStart = -1;
+            var segmentEnd = 0;
+
+            while (TryGetNextSegment(connectionString, segmentSeparator, ref segmentStart, ref segmentEnd))
+            {
+                var kvSeparatorIndex = connectionString.IndexOf(keyValueSeparator, segmentStart, segmentEnd - segmentStart, StringComparison.Ordinal);
+                int keyStart = GetStart(connectionString, segmentStart);
+                int keyLength = GetLength(connectionString, keyStart, kvSeparatorIndex);
+
+                var key = connectionString.Substring(keyStart, keyLength);
+                if (segments.ContainsKey(key))
+                {
+                    throw new InvalidOperationException($"Duplicated key '{key}'");
```

Comment:
I'm still concerned about using "key" in a public-facing exception.  In most of the connection string documentation, key is referring to an auth key.

Reply:
Renamed to `keyword`.